### PR TITLE
Prevent default built-in scripts from breaking minification groups. #9 #22

### PR DESCRIPTION
**Pull request by shadyvb**
It seems like WP default scripts don't get assigned a `group` value after all, the final argument of the call to register the admin-bar.js script is never used to assign the `group` data value, same goes for all other default scripts:

https://github.com/WordPress/WordPress/blob/master/wp-includes/script-loader.php#L327 :
https://github.com/WordPress/WordPress/blob/master/wp-includes/class.wp-dependencies.php#L127

unlike what wp_enqueue_script does:
`if ( $in_footer )  $wp_scripts->add_data( $_handle[0], 'group', 1 );`
https://github.com/WordPress/WordPress/blob/master/wp-includes/functions.wp-scripts.php#L157

So the solution here is to detect the `dep->args` value if no `extra['group']` is available, cause they're basically the same thing.

### CODE REVIEW

File: dependency-minification.php [hunk 1]
```diff
@@ -728,6 +728,9 @@ static function group_handles_by_extra( array $handles, WP_Dependencies $wp_deps
 			unset($extra['suffix']);
 			unset($extra['rtl']);
 			unset($extra['data']);
+			// Default scripts are not assigned 'group', so we use the original 'deps->args' value
+			if ( is_a( $wp_deps, 'WP_Scripts' ) && empty($extra['group']) && is_int($dep->args) )
+				$extra['group'] = $dep->args;
```
Comment (westonruter) [inline dependency-minification.php:733]: @shadyvb could you wrap this condition in braces? Even though PHP allows you to omit them if there is only one statement in a block, we try to always include them as it prevents mistakes down the road when new statements are added to a condition.

**Comment by westonruter** [comment]
@shadyvb wow, I'm impressed with your investigation here! Nice work! :+1: It looks like the scripts are getting combined optimally. However, I think the styles aren't:

``` html
<link rel='stylesheet' id='minified-1-css'  href='http://local.wordpress-trunk.dev/_minify/admin-bar.b9ba73ded18d5a4b2ab200f5e6a93e89.a8cf63fecab3eb3edf6760bc02f258a6.css' type='text/css' media='all' />
<link rel='stylesheet' id='twentythirteen-fonts-css'  href='//fonts.googleapis.com/css?family=Source+Sans+Pro%3A300%2C400%2C700%2C300italic%2C400italic%2C700italic%7CBitter%3A400%2C700&#038;subset=latin%2Clatin-ext' type='text/css' media='all' />
<link rel='stylesheet' id='minified-2-css'  href='http://local.wordpress-trunk.dev/_minify/genericons,twentythirteen-style.9c4bd324d02d9a42b0839adbca6ada68.b60b3d64d6c3c2d4d65e3367e945b403.css' type='text/css' media='all' />
<!--[if lt IE 9]>
<link rel='stylesheet' id='minified-3-css'  href='http://local.wordpress-trunk.dev/_minify/twentythirteen-ie.462b9261dc77b94ad45e0e5663d6edfd.0be26d23c2dd2249c93b53d50f7a557f.css' type='text/css' media='all' />
<![endif]-->
```

Could you investigate that too?

**Comment by shadyvb** [comment]
I don't think there is an issue with styles apart from what we have already in #23 regarding the scheme-less urls, the fonts style was added to the group successfully when i allowed external resources to be minifed AND added a proper scheme to the url.

I didn't check this in the code; but since the plugin considers the style hosted locally, and is unable to minify it, shouldn't this actually happen ? break up the group to be able to serve the erroneous file without minification ?
UPDATE: Actually this doesn't happen, when a file cannot be reached, the whole group is output without minification. Shouldn't we implement this ? because this might cause the plugin to lose lots of file in a single group because of one faulty resource.

**Comment by westonruter** [comment]
Oh, you're absolutely right. The font stylesheet probably just has to be
enqueued first so that it doesn't break up the local styles into two
separate bundles. By default like you said the external resources are
excluded by default. Thanks!

**Comment by westonruter** [comment]
Regarding:

> UPDATE: Actually this doesn't happen, when a file cannot be reached, the whole group is output without minification. Shouldn't we implement this ? because this might cause the plugin to lose lots of file in a single group because of one faulty resource.

So if an external resource fails to load, it should then create a bundle/group before and a bundle/group after? I suppose this would be useful, however the complexity added for this edge-case error condition may not be worth it. Perhaps a compromise would be to cache all external scripts indefinitely, and fallback to using them if an external script ever fails when the cache expires.

**Comment by westonruter** [comment]
Ah, regarding the CSS snippet I pasted in above, I didn't look closely enough. The dependency that didn't get concatenated was the one for the admin bar, as it gets enqueued before the external fonts request. So all I had to do was move the twentythirteen font enqueueing to the end, and then the admin-bar CSS got concatenated with the other local styles:

``` html
<link rel='stylesheet' id='minified-1-css'  href='http://local.wordpress-trunk.dev/_minify/admin-bar,genericons,twentythirteen-style.f5fe0a6db03f31aa25c9d8a438075efc.33a4c5a9ad1294c3f12b18e7f7e387f0.css' type='text/css' media='all' />
<!--[if lt IE 9]>
<link rel='stylesheet' id='minified-2-css'  href='http://local.wordpress-trunk.dev/_minify/twentythirteen-ie.462b9261dc77b94ad45e0e5663d6edfd.0be26d23c2dd2249c93b53d50f7a557f.css' type='text/css' media='all' />
<![endif]-->
<link rel='stylesheet' id='twentythirteen-fonts-css'  href='//fonts.googleapis.com/css?family=Source+Sans+Pro%3A300%2C400%2C700%2C300italic%2C400italic%2C700italic%7CBitter%3A400%2C700&#038;subset=latin%2Clatin-ext' type='text/css' media='all' />
```

**Comment by shadyvb** [comment]
Caching external scripts indefinitely is definitely :) a good idea!

However, the scenario where the cache of that faulty external url is never
created still exists, actually on my local installation with some
connectivity troubles, caching external enabled, and font style url
(twentythirteen) is schemed, depmin wasn't able to cache the font
stylesheet for some reason, which prevented the whole group from being
minified.

So i do think this is an issue to be considered after all, even with a low
priority, or at least mention it in the doc.

**Pull request merged**